src/graphics/env-lighting.js

```diff
@@ -0,0 +1,397 @@
+    //     });
+
+    //     // get the device to create gl interfaces
+    //     device.setRenderTarget(renderTarget);
+    //     device.updateBegin();
+
+    //     device.setTexture(target, 0);
+
+    //     // create the target mipmap level
+    //     gl.texImage2D(
+    //         glTarget,
+    //         level,
+    //         mipmap._glInternalFormat,
+    //         mipmap.width,
+    //         mipmap.height,
+    //         0,
+    //         mipmap._glFormat,
+    //         mipmap._glPixelType,
+    //         null
+    //     );
+
+    //     // copy it
+    //     gl.copyTexImage2D(
+    //         glTarget,
+    //         level,
+    //         mipmap._glInternalFormat,
+    //         0, 0, mipmap.width, mipmap.height,
+    //         0
+    //     );
+
+    //     device.updateEnd();
+    //     renderTarget.destroy();
+    // }
+
+    // // restore render target
+    // device.setRenderTarget(oldRt);
+    // device.updateBegin();
+};
+
+// generate mipmaps for the given target texture
+// target is either 2d equirect or cubemap with mipmaps = false
+const generateMipmaps = (target) => {
+    const device = target.device;
+    const gl = device.gl;
+    const numLevels = calcLevels(target.width, target.height);
+
+    let prevLevel = null;
+    for (let i = 1; i < numLevels; ++i) {
+        const level = new Texture(device, {
+            name: target.name + '-tmp-' + i,
+            cubemap: target.cubemap,
+            width: Math.max(1, Math.floor(target.width >> i)),
+            height: Math.max(1, Math.floor(target.height >> i)),
+            format: target.format,
+            type: target.type,
+            addressU: target.addressU,
+            addressV: target.addressV,
+            fixCubemapSeams: target.fixCubemapSeams,
+            mipmaps: false
+        });
+
+        // downsample texture using single sample reproject
+        reprojectTexture(prevLevel || target, level, {
+            numSamples: 1
+        });
+
+        // copy into target's mipmap
+        copyMipmap(target, level);
+
+        if (prevLevel) {
+            prevLevel.destroy();
+        }
+        prevLevel = level;
+    }
+
+    if (prevLevel) {
+        prevLevel.destroy();
+    }
+
+    // set filtering while bypassing the engine
+    device.setTexture(target, 0);
+    gl.texParameteri(target._glTarget, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
+    gl.texParameteri(target._glTarget, gl.TEXTURE_MIN_FILTER, gl.LINEAR_MIPMAP_LINEAR);
+    target._mipmaps = true;
+};
+
+// helper functions to support prefiltering lighting data
+class EnvLighting {
+    /**
+     * @private
+     * @function
+     * @name generateSkyboxCubemap
+     * @description Generate a skybox cubemap in the correct pixel format from the source texture.
+     * @param {Texture} source - The source texture. This is either a 2d texture in equirect format or a cubemap.
+     * @param {number} size - Size of the resulting texture. Specify 0 for automatic sizing.
```

Comment:
why not simply don't pass anything for automatic sizing

Reply:
The parameter is defaulted. I'll update the docs to tag it optional.